src/platform/cc13x2_26x2/KeyValueStoreManagerImpl.cpp

```diff
@@ -21,14 +21,68 @@
  *          Platform-specific key value storage implementation for CC1352
  */
 
+/* this file behaves like a config.h, comes first */
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
 #include <platform/KeyValueStoreManager.h>
 
+#include <platform/cc13x2_26x2/CC13X2_26X2Config.h>
+
 namespace chip {
 namespace DeviceLayer {
 namespace PersistedStorage {
 
+using namespace ::chip::DeviceLayer::Internal;
+
 KeyValueStoreManagerImpl KeyValueStoreManagerImpl::sInstance;
 
+/**
+ * Calculate crc16 CCITT from the c string key.
+ *
+ * This could be optimized because key is probably a printable string and not a
+ * binary array. The length is limited to 10 bits because the maximum value of
+ * an NV sub ID is 10 bits long. Consult
+ * `<simplelink_sdk>/source/ti/common/nv/nvocmp.c:NVOCMP_MAXSUBID`.
+ */
+static uint16_t calc_key_10(const char * key)
```

Comment:
This is clever and no doubt expedient for this platform.  I think we should all consider though that this type of implementation precludes ever implementing KVS iterators (get all key=value pairs) or filters (get all key=value pairs, where key matches PATTERN), unless we are willing to break compatibility with implementations like this to achieve such things.

It also seems like with only 10 bits to work with, users and stakeholders for this platform would benefit from some sort of unit test to check for likely key collisions.

As to the CRC algorithm itself, it might be worth specifying this more completely (polynomial, xor in and out, etc.).  Once this is deployed for this platform, it can't really be changed.  It would give some confidence that this is the CRC algorithm it claims to be if there was a pointer to some reputable snippet of source code, or if it could be shown that this produces the same output as e.g. source code from this code generator:

```
git clone https://github.com/tpircher/pycrc
python3 pycrc/pycrc.py --generate c -o other_crc_ccitt_impl.c --algorithm=bbf --model=crc-16-ccitt
```

Reply:
I'll definitely add that python reference, I didn't know it existed. This was the result of some googling for a CRC.

The underlying driver for this implementation can support iterators, we made it to support those goals when OpenThread asked for it. But those implementations knew what they were storing and knew if there were multiple instances of an item. That is information that can be known at compile time, but is obfuscated here by a string concatenation. It may be possible to do something like scanf to reconstruct that information at runtime.

This implementation is natural on a platform like Linux with a file system that can handle arbitrary paths to files. However, choices like this are going to freeze out existing embedded SoCs That do not have the space for a filesystem.

Reply:
I see.  Yes, you're right; if you know the possible key space, you can still iterate.  I hope that's sufficient, and can see why it's important to support this type of implementation.  And anyway, this implementation is conformant to the current kvstore API, so can't really be called on that account.

Glad you may potentially find pycrc useful.

> However, choices like this are going to freeze out existing embedded SoCs That do not have the space for a filesystem.

Things like fat12 are pretty micro friendly, but then you still only have 8.3 file names.  So still no solution for our key lengths.

Reply:
The FS we supply for internal storage in our SDK is spiffs. This increases the number of Flash pages for storage from 2 (16Kbytes) to 3 (24Kbytes) and adds the ~20 Kbytes of code increase for the software itself. Not to mention our BLE stack would have to be modified to use the file system, or some shim code would have to be added to support the existing storage API.

I will look into a way to store the keys in a parallel item ID, but that implementation will be pretty inefficient.